batch/models.py

```diff
@@ -82,6 +83,34 @@ def choices(cls):
         return [(key.value, key.name) for key in cls]
 
 
+class JobManager(models.Manager):
+    _STATUS_TRANSITIONS = {
+        (JobStatus.created, JobStatus.running): {"now": "started_on"},
+        (JobStatus.running, JobStatus.done): {"now": "finished_on"},
+        (JobStatus.running, JobStatus.failed): {"now": "finished_on"},
+        (JobStatus.done, JobStatus.collected): {},
+    }
+
+    def update_status(self, *, old: JobStatus, new: JobStatus) -> int:
+        """Update the job status of records with the old status to the new status.
+
+        Returns:
+            Number of changed rows.
+
+        Raises:
+            ValueError: Status transition is invalid.
+        """
+        transition_info = self._STATUS_TRANSITIONS.get((old, new))
+        if transition_info is None:
+            raise ValueError(f"invalid job status transition from {old.value} to {new.value}")
+
+        extra_update_fields = {}
+        if "now" in transition_info:
+            extra_update_fields[transition_info["now"]] = datetime.datetime.utcnow()
+
+        return self.filter(status=old).update(status=new, **extra_update_fields)
```

Comment:
Check returned value is 1. Given that we have id based lookup.

Reply:
Maybe we should have each transition as its own specific method, since they look like they might have transition-specific logic, like that `if "now" in transition info`. It might be easier to maintain and customize transitions that look like

```python3
class JobExecution:
    ...
    def transition_to_running(self):
        update_time = datetime.datetime.utcnow()
        num_records = self.filter(status=JobStatus.created).update(status=JobStatus.running, started_on=update_time)
        if num_records != 1:
            raise SomeException(...)
```

Reply:
> Maybe we should have each transition as its own specific method, since they look like they have transition-specific logic, like that `if "now" in transition info`

I don't see any value for this right now because the transition-specific logic is minimal. Also, transitions are driven by a status value anyway.